### ISSUE TITLE
yielding in ssl_certificate/preread yields a wrong light thread (parent thread)

### ISSUE THREAD

**Issue by bungle** [openresty/stream-lua-nginx-module#227]
Here is a test script:

```bash
#!/usr/bin/env bash
rm -f test/logs/error.log

mkdir -p test/conf
mkdir -p test/logs
mkdir -p test/ssl

openssl req -x509 -newkey rsa:2048 -keyout test/ssl/key.pem -out test/ssl/cert.pem -days 365 -nodes -subj "/C=US"

cat > test/conf/nginx.conf <<- CONF
worker_processes  1;
events {
    worker_connections  1024;
}
stream {
    server {
        listen  127.0.0.1:9000;
        listen  127.0.0.1:9443 ssl;
        ssl_certificate      ../ssl/cert.pem;
        ssl_certificate_key  ../ssl/key.pem;
        ssl_certificate_by_lua_block {
          ngx.log(ngx.ERR, "SSL CERTIFICATE...")
          local co = coroutine.create(function()
              ngx.log(ngx.ERR, "SLEEPING...")
              ngx.sleep(0.1)
              ngx.log(ngx.ERR, "SLEEPING DONE")
          end)
          ngx.log(ngx.ERR, "RESUMING COROUTINE...")
          local ok, err = coroutine.resume(co)
          if ok then
              ngx.log(ngx.ERR, "RESUMING COROUTINE DONE")
          else
              ngx.log(ngx.ERR, "RESUMING COROUTINE ERRORED: ", err)
          end
          ngx.log(ngx.ERR, "SSL CERTIFICATE DONE")
        }
        preread_by_lua_block {
            ngx.log(ngx.ERR, "PREREAD...")
            local co = coroutine.create(function()
                ngx.log(ngx.ERR, "SLEEPING...")
                ngx.sleep(0.1)
                ngx.log(ngx.ERR, "SLEEPING DONE")
            end)
            ngx.log(ngx.ERR, "RESUMING COROUTINE...")
            local ok, err = coroutine.resume(co)
            if ok then
                ngx.log(ngx.ERR, "RESUMING COROUTINE DONE")
            else
                ngx.log(ngx.ERR, "RESUMING COROUTINE ERRORED: ", err)
            end
            ngx.log(ngx.ERR, "PREREAD DONE")
        }
        content_by_lua_block {
          ngx.log(ngx.ERR, "CONTENT...")
          local co = coroutine.create(function()
              ngx.log(ngx.ERR, "SLEEPING...")
              ngx.sleep(0.1)
              ngx.log(ngx.ERR, "SLEEPING DONE")
          end)
          ngx.log(ngx.ERR, "RESUMING COROUTINE...")
          local ok, err = coroutine.resume(co)
          if ok then
              ngx.log(ngx.ERR, "RESUMING COROUTINE DONE")
          else
              ngx.log(ngx.ERR, "RESUMING COROUTINE ERRORED: ", err)
          end
          ngx.log(ngx.ERR, "CONTENT DONE")
        }
    }
}
CONF

openresty -p test
curl -s 127.0.0.1:9000
curl -s -k https://127.0.0.1:9443
openresty -p test -s stop

cat test/logs/error.log | grep error]
```

When I run this I get the following output:
```log
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] preread_by_lua(nginx.conf:42):2: PREREAD... while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] preread_by_lua(nginx.conf:42):8: RESUMING COROUTINE... while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] preread_by_lua(nginx.conf:42):4: SLEEPING... while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] preread_by_lua(nginx.conf:42):11: RESUMING COROUTINE DONE while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] preread_by_lua(nginx.conf:42):15: PREREAD DONE while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] content_by_lua(nginx.conf:58):2: CONTENT..., client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] content_by_lua(nginx.conf:58):8: RESUMING COROUTINE..., client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] content_by_lua(nginx.conf:58):4: SLEEPING..., client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] content_by_lua(nginx.conf:58):6: SLEEPING DONE, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] content_by_lua(nginx.conf:58):11: RESUMING COROUTINE DONE, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *1 stream [lua] content_by_lua(nginx.conf:58):15: CONTENT DONE, client: 127.0.0.1, server: 127.0.0.1:9000
2021/03/03 00:12:07 [error] 94626#0: *3 stream [lua] ssl_certificate_by_lua:2: SSL CERTIFICATE..., context: ssl_certificate_by_lua*, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *3 stream [lua] ssl_certificate_by_lua:8: RESUMING COROUTINE..., context: ssl_certificate_by_lua*, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *3 stream [lua] ssl_certificate_by_lua:4: SLEEPING..., context: ssl_certificate_by_lua*, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *3 stream [lua] ssl_certificate_by_lua:11: RESUMING COROUTINE DONE, context: ssl_certificate_by_lua*, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *3 stream [lua] ssl_certificate_by_lua:15: SSL CERTIFICATE DONE, context: ssl_certificate_by_lua*, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] preread_by_lua(nginx.conf:42):2: PREREAD... while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] preread_by_lua(nginx.conf:42):8: RESUMING COROUTINE... while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] preread_by_lua(nginx.conf:42):4: SLEEPING... while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] preread_by_lua(nginx.conf:42):11: RESUMING COROUTINE DONE while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] preread_by_lua(nginx.conf:42):15: PREREAD DONE while prereading client data, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] content_by_lua(nginx.conf:58):2: CONTENT..., client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] content_by_lua(nginx.conf:58):8: RESUMING COROUTINE..., client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] content_by_lua(nginx.conf:58):4: SLEEPING..., client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] content_by_lua(nginx.conf:58):6: SLEEPING DONE, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] content_by_lua(nginx.conf:58):11: RESUMING COROUTINE DONE, client: 127.0.0.1, server: 127.0.0.1:9443
2021/03/03 00:12:07 [error] 94626#0: *2 stream [lua] content_by_lua(nginx.conf:58):15: CONTENT DONE, client: 127.0.0.1, server: 127.0.0.1:9443
```

As can be seen the `sleep` in the `ssl_certificate_by_lua_block` and `preread_by_lua_block` phases yields its parent thread, the message `SLEEPING DONE` is not logged on those phases while it works correctly in `content_by_lua_block` phase.

**Comment by bungle** [comment]
This issue is at least with `1.19.3.1` and `1.17.8.2`, but pretty sure that it is in others too.

**Comment by bungle** [comment]
This seems to be fixed on latest `1.21.4.1`, thus closing it.

**Comment by bungle** [comment]
Thanks @catbro666 for fixing this with: https://github.com/openresty/stream-lua-nginx-module/pull/248
